packages/reflect-yjs/src/provider.ts

```diff
@@ -22,22 +32,73 @@ export class Provider {
     ydoc.on('update', this.#handleUpdate);
     ydoc.on('destroy', this.#handleDestroy);
 
-    this.#cancelSubscribe = reflect.subscribe<[string | null, string | null]>(
-      async tx => [
-        (await getServerUpdate(this.name, tx)) ?? null,
-        (await getClientUpdate(this.name, tx)) ?? null,
-      ],
-      ([serverUpdate, clientUpdate]) => {
-        if (serverUpdate !== null) {
-          this.#vector = Y.encodeStateVectorFromUpdateV2(
-            base64.toByteArray(serverUpdate),
-          );
+    const clientUpdateKey = yjsProviderClientUpdateKey(name);
+    const serverUpdateMetaKey = yjsProviderServerUpdateMetaKey(name);
+    const serverUpdateChunkKeyPrefix =
+      yjsProviderServerUpdateChunkKeyPrefix(name);
+
+    let isInitial = true;
+    this.#cancelWatch = reflect.experimentalWatch(
+      diff => {
+        let serverUpdateChange = false;
+        for (const diffOp of diff) {
+          const {key} = diffOp;
+          switch (diffOp.op) {
+            case 'add':
+            case 'change':
+              if (key === clientUpdateKey) {
+                this.#clientUpdate = base64.toByteArray(
+                  diffOp.newValue as string,
+                );
+              } else if (key === serverUpdateMetaKey) {
+                this.#serverUpdateMeta = diffOp.newValue as ChunkedUpdateMeta;
+                serverUpdateChange = true;
+              } else if (key.startsWith(serverUpdateChunkKeyPrefix)) {
+                this.#serverUpdateChunks.set(
+                  key.substring(serverUpdateChunkKeyPrefix.length),
+                  base64.toByteArray(diffOp.newValue as string),
+                );
+              }
+              break;
+            case 'del':
+              if (key === clientUpdateKey) {
+                this.#clientUpdate = null;
+              } else if (key === serverUpdateMetaKey) {
+                this.#serverUpdateMeta = null;
+                serverUpdateChange = true;
+              } else if (key.startsWith(serverUpdateChunkKeyPrefix)) {
+                this.#serverUpdateChunks.delete(
+                  key.substring(serverUpdateChunkKeyPrefix.length),
+                );
+              }
+              break;
+          }
+        }
+        if (serverUpdateChange) {
+          if (this.#serverUpdateMeta === null) {
+            this.#serverUpdate = null;
+            this.#vector = null;
+          } else {
+            this.#serverUpdate = unchunk(
```

Comment:
unfortunately this O(size-of-doc).

Reply:
I have ideas on how to optimize, see https://rocicorp.slack.com/archives/C013XFG80JC/p1702424077438199